Move show() to somewhere naturally inheritable / document what pyplot expects from a backend.

anntzer opened the pull request:
## PR Summary

~~Goes on top of #23090 (see https://github.com/matplotlib/matplotlib/pull/23090#issuecomment-1133763234).~~

### Move show() to somewhere naturally inheritable.
It's actually not clear whether to move it to FigureCanvas or
FigureManager.  FigureCanvas already has start_event_loop (cf.
start_main_loop), but pyplot_show/start_main_loop is more a
global/pyplot-only concept, so perhaps it belongs to FigureManager
instead?  OTOH, being on canvas_class makes it easier for switch_backend
to access it (it doesn't need to go through
`canvas_class.manager_class.pyplot_show`, which is relevant considering
that some designs of inheritable backends didn't even have a
manager_class attribute).

### Document what pyplot expects from a backend.

Finally, a document stating what's needed in a backend!
(only the pyplot interaction part, not the rendering part.)
Edit: now moved to #24218.

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->
**Tests and Styling**
- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of main, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

jklymak commented:
Failures looks real, and has picked up rebase issues...  This looks like a reasonable simplification, thought I'm not an expert

anntzer commented:
Thanks for pointing this out, this should be fixed now.  I would also like to request the reviewer's thoughts (perhaps @timhoffm?) as to whether pyplot_show should be a canvas or a manager method (as explained in the commit message/top post).

anntzer commented:
Discussed on the call: I'll move pyplot_show to the manager class instead, keeping that name; I'll also document explicitly that's it's only public for the purpose of being overridden by third-party backends, not for being called directly by end users.

anntzer commented:
The discussed modifications have been implemented.

timhoffm commented on lib/matplotlib/backend_bases.py:
```diff
@@ -3460,7 +3500,11 @@ def new_figure_manager_given_figure(cls, num, figure):
     def draw_if_interactive(cls):
-        if cls.mainloop is not None and is_interactive():
+        manager_class = cls.FigureCanvas.manager_class
+        interactive_backend = (
```
```suggestion
        backend_is_interactive = (
            # interactive backends need to overrwrite one of these methods
```

should we maybe have a `FigureManagerBase.has_interactive_backend()` that provides this functionality?

anntzer replied:
Changed the variable name.  I'd rather not add a method as the relevant customization point has been changed by #23291 anyways.

anntzer commented:
I have addressed your comments (thanks for the review) but moved the docs part of the PR to #24218, to make the pure implementation part easier to merge.

tacaswell commented on lib/matplotlib/backend_bases.py:
```diff
@@ -2822,6 +2822,46 @@ def create_with_canvas(cls, canvas_class, figure, num):
+                manager.show()  # Emits a warning for non-interactive backend.
+            except NonGuiException as exc:
+                _api.warn_external(str(exc))
+        if block is None:
```
Do we also want to check if pypplot is already imported?

tacaswell commented on lib/matplotlib/backends/backend_qt.py:
```diff
@@ -585,6 +585,12 @@ def resize(self, width, height):
 
+    @classmethod
+    def start_main_loop(cls):
+        qapp = QtWidgets.QApplication.instance()
```
We now do the "does the app exist" dance in `__init__`, but this is a static method so do we want to use `_creatApp` here as well?

tacaswell replied:
or at least bail if it does not exist.

anntzer replied:
sure

tacaswell commented on lib/matplotlib/backends/backend_webagg.py:
```diff
@@ -307,21 +325,3 @@ def ipython_inline_display(figure):
 class _BackendWebAgg(_Backend):
     FigureCanvas = FigureCanvasWebAgg
     FigureManager = FigureManagerWebAgg
-
```
missing the re-export?

tacaswell replied:
nvm, crossed this with the mainloop changes.

tacaswell commented on lib/matplotlib/backends/_backend_tk.py:
```diff
@@ -486,6 +486,20 @@ def create_with_canvas(cls, canvas_class, figure, num):
+        managers = Gcf.get_all_fig_managers()
+        if managers:
+            first_manager = managers[0]
+            manager_class = type(first_manager)
```
```suggestion
            manager_class = cls
```

anntzer replied:
this is copy-pasted from the previous implementation; I do not feel comfortable changing that.

timhoffm commented on lib/matplotlib/backend_bases.py:
```diff
@@ -2840,6 +2840,48 @@ def create_with_canvas(cls, canvas_class, figure, num):
+        """
+        Start the main event loop.
+
+        Interactive backends need to reimplement this method or
+        `~.FigureManagerBase.pyplot_show`.
```
Can you be more specific? What is the criterion to `start_main_loop()` vs. `pyplot_show()`?

anntzer replied:
Explained a bit more.

timhoffm commented on lib/matplotlib/backend_bases.py:
```diff
@@ -3518,7 +3560,11 @@ def new_figure_manager_given_figure(cls, num, figure):
-        if cls.mainloop is not None and is_interactive():
+        manager_class = cls.FigureCanvas.manager_class
+        backend_is_interactive = (
+            manager_class.start_main_loop != FigureManagerBase.start_main_loop
```
```suggestion
            # interactive backends reimplement either start_main_loop or pyplot_show
            manager_class.start_main_loop != FigureManagerBase.start_main_loop
```
Or should we put this into a class method so that we can do
```
manager_class.backend_is_interactive()
```
?

anntzer replied:
I'd rather not.  First, this is basically the definition of interactiveness that matches the old behavior of draw_if_interactive (overriding start_main_loop or pyplot_show is essentially like overriding mainloop previously), but there's also a different definition, namely that manager.show doesn't raise a NonGuiException.  In practice, classes that override manager.show to not raise a NonGuiException are also those that override s_m_l/p_s, but I don't want to formalize this definition, in particular, because overriding s_m_l/p_s also only matters for draw_if_interactive, which is becoming "semi-deprecated" since #23291.

timhoffm commented on lib/matplotlib/backend_bases.py:
```diff
@@ -3546,8 +3592,8 @@ def show(cls, *, block=None):
             # Hack: Are we in IPython's %pylab mode?  In pylab mode, IPython
             # (>= 0.10) tacks a _needmain attribute onto pyplot.show (always
             # set to False).
-            from matplotlib import pyplot
-            ipython_pylab = hasattr(pyplot.show, "_needmain")
+            ipython_pylab = hasattr(
+                getattr(sys.modules.get("pyplot"), "show", None), "_needmain")
```
Move this to a helper function in `cbook`?

anntzer replied:
The duplication is only temporary, because the version in _Backend.show will go away as I plan to get rid of the _Backend class, which was an implementation detail anyways.  (I'm not doing it immediately because the ipympl backend has started relying on it, but I plan to help them to get rid of that use, either going back to the old-style fully manual API if they want a long backcompat, or going to the new class-based API.)
